rfc: RFC 17 Guidance on the use of LLMS (#435)
thousandbrainsproject/tbp.monty: rfcs/0000_guidance_on_using_llms.md

- LLM code is often verbose, or results in excessive, unnecessary changes to the code.

Due to the above reasons, LLM-generated code can take a great deal of time to review and debug. This can be avoided when PRs are written with intent by a person.

### Legal Concerns
- There are non-trivial legal concerns around contamination of code when using LLMs. See for example, [this recent study](https://arxiv.org/html/2408.02487v1), which demonstrates that LLM-generated code can violate copyright restrictions in a significant number of cases.
- A cornerstone of the Thousand Brains Project is its open source nature, and this motivates our use of an MIT licence when we distribute our code. However, the inadvertent integration of copyright-protected code into `thousandbrainsproject/tbp.monty` could jeopardize the ability to make the code open-source, disrupting any distributions of the platform.

### Take-Aways

The high-level guidance based on the above is:
- Using an LLM to auto-complete variable names and other simple speed-ups can be appropriate.

> - Multi-line sections of algorithmic code written by LLMs should be thoroughly checked for logical correctness and potential copyright violations before opening a PR into `thousandbrainsproject/tbp.monty`.

Author: thought: It would be nice if we could provide some concrete guidance on how to check for legal compliance. @codeallthethingz did you have any more luck looking into this?

LLM code-assistants have the potential to improve the productivity of coders. More generally, there is optimism that AI tools might accelerate science. At the same time, their use can introduce non-intuitive drawbacks. The aim of this RFC is to describe these in more detail, and agree on what guidance we should ask of both ourselves and the community.

# Guidance on the Use of AI/LLMs

## High-Level Guidance
The below sections provide descriptions of concerning attributes associated with LLM-generated contributions, and why these can be problematic. While we do not institute any strict guidance, please note: if we feel that the use of AI/LLMs has resulted in a contribution which matches these concerning patterns, Maintainers reserve the right to not review your contribution. We do not want to discourage contributions, but this balancing act on our part is an unfortunate side-effect of the small size of the Maintainers team vs. the speed at which LLM content can be generated.

## Writing RFCs

We suggest that all RFCs are written by you, and that you try to avoid using Large Language Models (LLMs) like ChatGPT in the writing process. Some of our motivations for this are:

> - We want to get a clear understanding of your solution, and we have found that LLMs give poor, if approximately correct, RFC proposals. This is particularly the case because the work of the TBP falls very much in the "out of training distribution" domain.

Reviewer: ```suggestion
- We want to get a clear understanding of your solution, and we have found that LLMs give poor, if approximately correct, RFC proposals. The approximate correctness makes it more difficult to spot logical inconsistencies, and the additional verbosity makes them time-consuming to review. Logical inconsistencies are particularly common because the work of the TBP falls very much in the "out of training distribution" domain.
```

- When we review RFCs written by LLMs, it often ends up taking more time both for us, and for you, so it is much better if you write them yourself.

> 

Reviewer: ```suggestion
- Writing the RFC yourself will serve as a useful starting point if you intend to implement the idea afterwards. Formulating the proposal in your own words often surfaces issues and ideas you didn't think about previously, which is not something that usually happens when you have an LLM generate the proposal for you. Note that RFCs don't need to be long-winded arguments and detailed elaborations of every step you intend to take. They are meant to be a concise summary of your idea and form the basis for discussion and feedback from others. As such, it brings you no benefits, both in your time spent and understanding gained, if they are LLM-generated.
```

Maybe something along those lines could be worth adding?

In principle, you may use LLMs, such as code-assistants, when writing code that you contribute to the Thousand Brains Project. However, we suggest that you do so in a limited manner, being mindful of the below issues:

### Quality Concerns
- As with RFCs, we have found that our code-base is out-of-distribution, and the quality of code written by LLMs, while superficially correct, often is not.
- Similarly, Monty is fundamentally about developing sensorimotor AI that can succeed in embodied settings. These are precisely the settings that LLMs struggle to perform in, and so once again we have found they frequently provide incorrect solutions.
- LLM code is often verbose, or results in excessive, unnecessary changes to the code.

Due to the above reasons, LLM-generated code can take a great deal of time to review and debug. This can be avoided when PRs are written with intent by a person.

### Legal Concerns
- There are non-trivial legal concerns around contamination of code when using LLMs. See for example, [this recent study](https://arxiv.org/html/2408.02487v1), which demonstrates that LLM-generated code can violate copyright restrictions in a significant number of cases.

> - A cornerstone of the Thousand Brains Project is its open source nature, and this motivates our use of an MIT licence when we distribute our code. However, the inadvertent integration of copyright-protected code into `thousandbrainsproject/tbp.monty` could jeopardize the ability to make the code open-source, disrupting any distributions of the platform.

Reviewer: ```suggestion
- A cornerstone of the Thousand Brains Project is its open-source nature, and this motivates our use of an MIT licence when we distribute our code. However, the inadvertent integration of copyright-protected code into `thousandbrainsproject/tbp.monty` could jeopardize the ability to make the code open-source, disrupting any distributions of the platform.
```

### Take-Aways

The high-level guidance based on the above is:
- Using an LLM to auto-complete variable names and other simple speed-ups is appropriate.
- Multi-line sections of algorithmic code written by LLMs should be checked for logical correctness and potential copyright violations before opening a PR into `thousandbrainsproject/tbp.monty`.

Below we provide further guidance on some edge cases.

#### Work on Research Prototypes

> - [Research Prototypes](https://github.com/nielsleadholm/tbp.monty/blob/978b15653a4c08bb21e28752a2ea9e01a3da906b/rfcs/0000_code_guidance_for_researchers_and_community.md) are separate forks of `thousandbrainsproject/tbp.monty` intended to rapidly evaluate the merits of a particular research idea. As they are not part of the core Monty platform, the legal concerns described above are less relevant, however the code-quality issues remain.

Reviewer 2: **issue**: The `Research Prototypes` link will break when that RFC is numbered. 

**suggestion**: I recommend linking to the Pull Request for non-merged RFCs, which will always be stable.

- On the other hand, we suggest that you do not pass a [Future Work item](https://thousandbrainsproject.readme.io/docs/project-roadmap) description into an LLM, and then open a PR with all of the code it generated.

## Contributing on the Forums

- [The TBP Discourse forums](https://thousandbrains.discourse.group/) are an excellent setting to discuss nascent ideas.
- In these discussions, we would love to engage with you and your ideas. As such, we ask that you not post the outputs of LLMs as if these were your own ideas.
- If you would like to quote or refer to a conversation you have had with an LLM, please just indicate this clearly.
- See the advice under [Writing RFCs](#writing-rfcs) for suggestions about when LLMs can be useful "tutors" when exploring a topic.

## Assistive Technology

> - Please note that the above guidance does not apply to assistive technology such as grammar checkers, which we understand you might find helpful in copy-editing text in RFCs and forum posts. If you need to use a more advanced AI tool for a specific reason, we ask that you reach out to us directly to discuss this beforehand.

Reviewer 2: **suggestion**: I would rephrase the last bit to be less restrictive: "..., we ask that you keep the concerns in this RFC in mind."

**issue**: I think "assistive" is associated with accessibility and has possibly legal implications.

**suggestion**: Maybe use a different term, "Productivity Technology"?

# Motivation

We're really excited to receive contributions from the wider community - one of the highlights of our day is when we see a new PR or RFC open from a non-Maintainer.

The reason for this guide is that we have experienced a few instances of suspected LLM-based contributions. In these cases, the use of LLMs in writing material became an obstacle between us working together with the proposer, resulting in issues including long review times.

As such, the below sections provide descriptions of the kinds of concerning attributes that are associated with LLM-generated contributions, and why these can be problematic. While we do not institute any strict guidance, please note that if we feel that the use of AI/LLMs has resulted in a contribution that matches these concerning patterns, Maintainers reserve the right to not review your contribution. We do not want to discourage contributions, but this balancing act on our part is an unfortunate side-effect of the small size of the Maintainers team vs. the speed at which LLM content can be generated.

# Guidance on the Use of AI/LLMs

## High-Level Guidance

> The below sections provide descriptions of concerning attributes associated with LLM-generated contributions, and why these can be problematic. While we do not institute any strict guidance, please note: if we feel that the use of AI/LLMs has resulted in a contribution which matches these concerning patterns, Maintainers reserve the right to not review your contribution. We do not want to discourage contributions, but this balancing act on our part is an unfortunate side-effect of the small size of the Maintainers team vs. the speed at which LLM content can be generated.

Reviewer 3: I think a large chunk of this paragraph is repeated from the previous paragraph.

# Motivation

We're really excited to receive contributions from the wider community - one of the highlights of our day is when we see a new PR or RFC open from a non-Maintainer.

> The reason for this guide is that we have experienced a few instances of suspected LLM-based contributions. In these cases, the use of LLMs in writing material became an obstacle between us working together with the proposer, resulting in issues including long review times.

Reviewer: ```suggestion
The reason for this guide is that we are increasingly encountering the limits of LLMs and how those can negatively affect our time spent on something and the quality of the output.
```
I wouldn't call this out specifically. We wouldn't want anyone to feel bad who made an LLM based contribution
